**Linear Regression**

Feature Engineering:

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

df = pd.read_csv('spotify_music_cleaned.csv')
df.shape


(89583, 20)

In [8]:
#one-hot encoding track_genre so all genres are weighted the same instead of being read as numerical rankings
df = pd.get_dummies(df, columns=['track_genre'])

df.shape
#df.columns.tolist()


(89583, 132)

In [9]:
y = df['popularity']
X = df.drop(columns=['popularity', 'track_id', 'artists', 'track_name', 'album_name', ])

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 27)

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(71666, 127)
(17917, 127)
(71666,)
(17917,)


In [11]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [12]:
model = LinearRegression()

model.fit(X_train_scaled, y_train)
predictions = model.predict(X_test_scaled)

/Users/jalentran/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/jalentran/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/jalentran/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_


In [13]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_test, predictions)
r2 = r2_score(y_test, predictions)
rmse = mean_squared_error(y_test, predictions) ** 0.5

print(mae, r2, rmse)

12.07029977294915 0.32542737636407393 16.90526150000689


**Observations after training**

- The Mean Absolute Error score was 12.07, meaning my model was off by about 12 popularity points from the correct answer on average (using the popularity scale 0-100)

- The R squared value the model got was 0.325 meaning about 32.5% of the variation in popularity across songs can be predicted by the linear regression model.

- Lastly the Root Mean Squared Error value was 16.905 meaning there was not very consistent error sizes across the board. There were some higher error values.

- This was somewhat the results I was expecting. After visualizing the data earlier, I noticed there was no consistent linear pattern for the relationship between variables and song popularity. The r value (correlation coefficient) was a little higher than expected. 